core/ide/che-core-ide-api/src/main/java/org/eclipse/che/ide/api/project/MutableProjectConfig.java

```diff
@@ -0,0 +1,172 @@
+        name = source.getName();
+        path = source.getPath();
+        description = source.getDescription();
+        type = source.getType();
+        mixins = newArrayList(source.getMixins());
+        attributes = newHashMap(source.getAttributes());
+        sourceStorage = new MutableSourceStorage(source.getSource());
+    }
+
+    public MutableProjectConfig() {
+    }
+
+    @Override
+    public String getName() {
+        return name;
+    }
+
+    public void setName(String name) {
+        this.name = name;
+    }
+
+    @Override
+    public String getPath() {
+        return path;
+    }
+
+    public void setPath(String path) {
+        this.path = path;
+    }
+
+    @Override
+    public String getDescription() {
+        return description;
+    }
+
+    public void setDescription(String description) {
+        this.description = description;
+    }
+
+    @Override
+    public String getType() {
+        return type;
+    }
+
+    public void setType(String type) {
+        this.type = type;
+    }
+
+    @Override
+    public List<String> getMixins() {
+        if (mixins == null) {
+            mixins = new ArrayList<>();
+        }
+
+        return mixins;
+    }
+
+    public void setMixins(List<String> mixins) {
+        this.mixins = mixins;
+    }
+
+    @Override
+    public Map<String, List<String>> getAttributes() {
+        if (attributes == null) {
+            attributes = newHashMap();
+        }
+
+        return attributes;
+    }
+
+    public void setAttributes(Map<String, List<String>> attributes) {
+        this.attributes = attributes;
+    }
+
+    @Override
+    public MutableSourceStorage getSource() {
+        if (sourceStorage == null) {
+            sourceStorage = new MutableSourceStorage();
+        }
+
+        return sourceStorage;
+    }
+
+    public void setSource(SourceStorage sourceStorage) {
+        this.sourceStorage = new MutableSourceStorage(sourceStorage);
+    }
+
+    public class MutableSourceStorage implements SourceStorage {
+        private String              type;
+        private String              location;
+        private Map<String, String> parameters;
+
+        public MutableSourceStorage(SourceStorage source) {
+            type = source.getType();
+            location = source.getLocation();
+            parameters = source.getParameters();
+        }
+
+        public MutableSourceStorage() {
+        }
+
+        @Override
+        public String getType() {
+            return type;
+        }
+
+        public void setType(String type) {
+            this.type = type;
+        }
+
+        @Override
+        public String getLocation() {
+            return location;
+        }
+
+        public void setLocation(String location) {
+            this.location = location;
+        }
+
+        @Override
+        public Map<String, String> getParameters() {
+            if (parameters == null) {
+                parameters = newHashMap();
```

Comment:
`java.util.Collections#emptyMap` returns an immutable map which is not suitable in this case. The point of `MutableProjectConfig` is to allow modify configuration.